File: packages/typescript/src/character.ts
import { validateCharacter } from "./schemas/character";
import type {
	Character,
	CharacterSettings,
	KnowledgeSourceItem,
	MessageExample,
	MessageExampleGroup,
} from "./types";

type LegacyKnowledgeItem =
	| string
	| { path: string; shared?: boolean }
	| { directory: string; shared?: boolean }
	| KnowledgeSourceItem;

type MessageExamplesInput = MessageExampleGroup[] | MessageExample[][];

export interface CharacterInput {
	id?: string;
	name?: string;
	username?: string;
	system?: string;
	templates?: Record<string, string>;
	bio?: string | string[];
	messageExamples?: MessageExamplesInput;
	postExamples?: string[];
	topics?: string[];
	adjectives?: string[];
	knowledge?: LegacyKnowledgeItem[];
	plugins?: string[];
	settings?: CharacterSettings;
	secrets?: Record<string, string>;
	style?: { all?: string[]; chat?: string[]; post?: string[] };
	advancedPlanning?: boolean;
	advancedMemory?: boolean;
}

interface NormalizedCharacterInput {
	id?: string;
	name?: string;
	username?: string;
	system?: string;
	templates: Record<string, string>;
	bio: string[];
	messageExamples: MessageExampleGroup[];
	postExamples: string[];
	topics: string[];
	adjectives: string[];
	knowledge: KnowledgeSourceItem[];
	plugins: string[];
	settings?: CharacterSettings;
	secrets: Record<string, string>;
	style?: { all?: string[]; chat?: string[]; post?: string[] };
	advancedPlanning?: boolean;
	advancedMemory?: boolean;
}

const isRecord = (value: unknown): value is Record<string, unknown> =>
	typeof value === "object" && value !== null;

const isMessageExampleGroup = (
	value: MessageExampleGroup | MessageExample[],
): value is MessageExampleGroup =>
	isRecord(value) && "examples" in value && Array.isArray(value.examples);

function normalizeMessageExamples(
	input?: MessageExamplesInput,
): MessageExampleGroup[] {
	if (!input || input.length === 0) return [];
	const first = input[0];
	if (Array.isArray(first)) {
		const exampleSets = input as MessageExample[][];
		return exampleSets.map((examples) => ({ examples }));
	}
	if (isMessageExampleGroup(first)) {
		return input as MessageExampleGroup[];
	}
	return [];
}

function normalizeKnowledgeItem(
	item: LegacyKnowledgeItem,
): KnowledgeSourceItem | null {
	if (typeof item === "string") {
		return { item: { case: "path", value: item } };
	}
	if (!isRecord(item)) {
		return null;
	}
	if ("item" in item && isRecord(item.item)) {
		const caseValue = item.item.case;
		if (caseValue === "path" && typeof item.item.value === "string") {
			return item as KnowledgeSourceItem;
		}
		if (
			caseValue === "directory" &&
			isRecord(item.item.value) &&
			typeof item.item.value.path === "string"
		) {
			return item as KnowledgeSourceItem;
		}
	}
	if ("path" in item && typeof item.path === "string") {
		return { item: { case: "path", value: item.path } };
	}
	if ("directory" in item && typeof item.directory === "string") {
		return {
			item: {
				case: "directory",
				value: {
					directory: item.directory,
					shared: typeof item.shared === "boolean" ? item.shared : undefined,
				},
			},
		};
	}
	return null;
}

export function normalizeCharacterInput(
	input: CharacterInput,
): NormalizedCharacterInput {
	const bioValue = input.bio;
	const normalizedBio =
		bioValue === undefined
			? []
			: Array.isArray(bioValue)
				? bioValue
				: [bioValue];

	const normalizedKnowledge = (input.knowledge ?? [])
		.map((item) => normalizeKnowledgeItem(item))
		.filter((item): item is KnowledgeSourceItem => item !== null);

	return {
		id: input.id,
		name: input.name,
		username: input.username,
		system: input.system,
		templates: input.templates ?? {},
		bio: normalizedBio,
		messageExamples: normalizeMessageExamples(input.messageExamples),
		postExamples: input.postExamples ?? [],
		topics: input.topics ?? [],
		adjectives: input.adjectives ?? [],
		knowledge: normalizedKnowledge,
		plugins: input.plugins ?? [],
		settings: input.settings ?? {},
		secrets: input.secrets ?? {},
		style: input.style,
		advancedPlanning: input.advancedPlanning,
		advancedMemory: input.advancedMemory,
	};
}

export function createCharacter(
	input: CharacterInput & { name: string },
): Character {
	return normalizeCharacterInput(input) as Character;
}

export function parseCharacter(
	input: string | object | Character | CharacterInput,
): Character {
	if (typeof input === "string") {
		throw new Error(
			`Character path provided but must be loaded first: ${input}`,
		);
	}

	if (typeof input === "object") {
		const normalized =
			input && typeof input === "object"
				? normalizeCharacterInput(input as CharacterInput)
				: input;
		const validationResult = validateCharacter(normalized);

		if (!validationResult.success) {
			const validationError = validationResult.error;
			const errorDetails = validationError?.issues
				? validationError.issues
						.map((issue) => `${issue.path.join(".")}: ${issue.message}`)
						.join("; ")
				: validationError?.message || "Unknown validation error";
			throw new Error(`Character validation failed: ${errorDetails}`);
		}

		return validationResult.data as Character;
	}

	throw new Error("Invalid character input format");
}

export function validateCharacterConfig(character: Character): {
	isValid: boolean;
	errors: string[];
} {
	const validationResult = validateCharacter(character);

	if (validationResult.success) {
		return {
			isValid: true,
			errors: [],
		};
	}

	const validationError = validationResult.error;
	const errors = validationError?.issues
		? validationError.issues.map(
				(issue) => `${issue.path.join(".")}: ${issue.message}`,
			)
		: [validationError?.message || "Unknown validation error"];

	return {
		isValid: false,
		errors,
	};
}

export function mergeCharacterDefaults(char: CharacterInput): Character {
	const normalized = normalizeCharacterInput(char);
	return {
		...normalized,
		name: normalized.name || "Unnamed Character",
	} as Character;
}

export function buildCharacterPlugins(
	env: Record<string, string | undefined> = process.env,
): string[] {
	const plugins = [
		"@elizaos/plugin-sql",
		...(env.ANTHROPIC_API_KEY?.trim() ? ["@elizaos/plugin-anthropic"] : []),
		...(env.OPENROUTER_API_KEY?.trim() ? ["@elizaos/plugin-openrouter"] : []),
		...(env.OPENAI_API_KEY?.trim() ? ["@elizaos/plugin-openai"] : []),
		...(env.GOOGLE_GENERATIVE_AI_API_KEY?.trim()
			? ["@elizaos/plugin-google-genai"]
			: []),
		...(env.DISCORD_API_TOKEN?.trim() ? ["@elizaos/plugin-discord"] : []),
		...(env.X_API_KEY?.trim() &&
		env.X_API_SECRET?.trim() &&
		env.X_ACCESS_TOKEN?.trim() &&
		env.X_ACCESS_TOKEN_SECRET?.trim()
			? ["@elizaos/plugin-x"]
			: []),
		...(env.TELEGRAM_BOT_TOKEN?.trim() ? ["@elizaos/plugin-telegram"] : []),
		...(!env.ANTHROPIC_API_KEY?.trim() &&
		!env.OPENROUTER_API_KEY?.trim() &&
		!env.OPENAI_API_KEY?.trim() &&
		!env.GOOGLE_GENERATIVE_AI_API_KEY?.trim()
			? ["@elizaos/plugin-ollama"]
			: []),
	];

	return plugins;
}

File: packages/rust/src/character.rs
//! Character parsing and validation for elizaOS
//!
//! This module provides functions for loading, parsing, and validating character files.

use crate::types::agent::{Character, CharacterSettings};
use anyhow::{Context, Result};
use std::collections::HashMap;

/// Parse a character from a JSON string
pub fn parse_character(json: &str) -> Result<Character> {
    let character: Character =
        serde_json::from_str(json).context("Failed to parse character JSON")?;
    validate_character(&character).context("Character validation failed")?;
    Ok(character)
}

/// Validate a character configuration
pub fn validate_character(character: &Character) -> Result<()> {
    // Validate name
    if character.name.is_empty() {
        anyhow::bail!("Character name is required");
    }

    // Validate plugins if present
    if let Some(plugins) = &character.plugins {
        for plugin in plugins {
            if plugin.is_empty() {
                anyhow::bail!("Empty plugin name in plugins list");
            }
        }
    }

    Ok(())
}

/// Merge character with default values
pub fn merge_character_defaults(mut character: Character) -> Character {
    // Apply defaults
    if character.settings.is_none() {
        character.settings = Some(CharacterSettings::default());
    }

    if character.plugins.is_none() {
        character.plugins = Some(vec![]);
    }

    // Ensure name matches TS/Py defaults (empty string -> "Unnamed Character")
    if character.name.is_empty() {
        character.name = "Unnamed Character".to_string();
    }

    character
}

/// Build character plugins based on environment variables
pub fn build_character_plugins(env: &HashMap<String, String>) -> Vec<String> {
    let mut plugins: Vec<String> = vec!["@elizaos/plugin-sql".to_string()];

    // Text-only plugins (no embedding support)
    if env
        .get("ANTHROPIC_API_KEY")
        .map(|s| !s.trim().is_empty())
        .unwrap_or(false)
    {
        plugins.push("@elizaos/plugin-anthropic".to_string());
    }
    if env
        .get("OPENROUTER_API_KEY")
        .map(|s| !s.trim().is_empty())
        .unwrap_or(false)
    {
        plugins.push("@elizaos/plugin-openrouter".to_string());
    }

    // Embedding-capable plugins (before platform plugins)
    if env
        .get("OPENAI_API_KEY")
        .map(|s| !s.trim().is_empty())
        .unwrap_or(false)
    {
        plugins.push("@elizaos/plugin-openai".to_string());
    }
    if env
        .get("GOOGLE_GENERATIVE_AI_API_KEY")
        .map(|s| !s.trim().is_empty())
        .unwrap_or(false)
    {
        plugins.push("@elizaos/plugin-google-genai".to_string());
    }

    // Platform plugins
    if env
        .get("DISCORD_API_TOKEN")
        .map(|s| !s.trim().is_empty())
        .unwrap_or(false)
    {
        plugins.push("@elizaos/plugin-discord".to_string());
    }
    if has_x_config(env) {
        plugins.push("@elizaos/plugin-x".to_string());
    }
    if env
        .get("TELEGRAM_BOT_TOKEN")
        .map(|s| !s.trim().is_empty())
        .unwrap_or(false)
    {
        plugins.push("@elizaos/plugin-telegram".to_string());
    }

    // BasicCapabilities plugin is now part of @elizaos/core and loaded automatically
    // No need to explicitly add it to the plugins list

    // Ollama fallback (only if no other LLM providers configured)
    let has_llm_provider = env
        .get("ANTHROPIC_API_KEY")
        .map(|s| !s.trim().is_empty())
        .unwrap_or(false)
        || env
            .get("OPENROUTER_API_KEY")
            .map(|s| !s.trim().is_empty())
            .unwrap_or(false)
        || env
            .get("OPENAI_API_KEY")
            .map(|s| !s.trim().is_empty())
            .unwrap_or(false)
        || env
            .get("GOOGLE_GENERATIVE_AI_API_KEY")
            .map(|s| !s.trim().is_empty())
            .unwrap_or(false);

    if !has_llm_provider {
        plugins.push("@elizaos/plugin-ollama".to_string());
    }

    plugins
}

fn has_x_config(env: &HashMap<String, String>) -> bool {
    env.get("X_API_KEY")
        .map(|s| !s.trim().is_empty())
        .unwrap_or(false)
        && env
            .get("X_API_SECRET")
            .map(|s| !s.trim().is_empty())
            .unwrap_or(false)
        && env
            .get("X_ACCESS_TOKEN")
            .map(|s| !s.trim().is_empty())
            .unwrap_or(false)
        && env
            .get("X_ACCESS_TOKEN_SECRET")
            .map(|s| !s.trim().is_empty())
            .unwrap_or(false)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::agent::Bio;

    #[test]
    fn test_parse_character_basic() {
        let json = r#"{
            "name": "TestAgent",
            "bio": "A test agent for testing purposes"
        }"#;

        let character = parse_character(json).unwrap();
        assert_eq!(character.name, "TestAgent");
    }

    #[test]
    fn test_parse_character_with_array_bio() {
        let json = r#"{
            "name": "TestAgent",
            "bio": ["Line 1", "Line 2"]
        }"#;

        let character = parse_character(json).unwrap();
        assert_eq!(character.name, "TestAgent");
        match &character.bio {
            Bio::Multiple(v) => assert_eq!(v.len(), 2),
            _ => panic!("Expected multiple bio"),
        }
    }

    #[test]
    fn test_validate_character() {
        let character = Character {
            name: "TestAgent".to_string(),
            bio: Bio::Single("A test agent".to_string()),
            ..Default::default()
        };

        assert!(validate_character(&character).is_ok());
    }

    #[test]
    fn test_validate_character_empty_name() {
        let character = Character {
            name: "".to_string(),
            bio: Bio::Single("A test agent".to_string()),
            ..Default::default()
        };

        assert!(validate_character(&character).is_err());
    }

    #[test]
    fn test_merge_character_defaults() {
        let character = Character {
            name: "TestAgent".to_string(),
            bio: Bio::Single("".to_string()),
            ..Default::default()
        };

        let merged = merge_character_defaults(character);
        assert!(merged.settings.is_some());
        assert!(merged.plugins.is_some());
    }

    #[test]
    fn test_build_character_plugins_empty_env() {
        let env = HashMap::new();
        let plugins = build_character_plugins(&env);

        // Should include sql and ollama fallback; basic_capabilities is now part of core
        assert!(plugins.contains(&"@elizaos/plugin-sql".to_string()));
        assert!(plugins.contains(&"@elizaos/plugin-ollama".to_string()));
    }

    #[test]
    fn test_build_character_plugins_with_openai() {
        let mut env = HashMap::new();
        env.insert("OPENAI_API_KEY".to_string(), "test-key".to_string());

        let plugins = build_character_plugins(&env);

        assert!(plugins.contains(&"@elizaos/plugin-openai".to_string()));
        assert!(!plugins.contains(&"@elizaos/plugin-ollama".to_string())); // No fallback
    }
}
